library.c

```diff
@@ -1108,6 +1108,17 @@ int redis_cmd_append_sstr_key_zstr(smart_string *dst, zend_string *key, RedisSoc
     return redis_cmd_append_sstr_key(dst, ZSTR_VAL(key), ZSTR_LEN(key), redis_sock, slot);
 }
 
+int redis_cmd_append_sstr_key_zval(smart_string *dst, zval *zv, RedisSock *redis_sock, short *slot) {
+    zend_string *key;
+    int res;
+
+    key = zval_get_string(zv);
+    res = redis_cmd_append_sstr_key_zstr(dst, key, redis_sock, slot);
```

Comment:
Can we use `Z_STR_P` here instead of `zval_get_string` + `zend_string_release`?

Reply:
The `zval` may not be a string.  It could be any other scalar value, or even an object that implements `__toString()`.

```php
<?php

class foo {
    public function __toString() {
        return 'zs1';
    }
}

$r = new Redis; $r->connect('localhost', 6379);
$r->del('zs1', 'zs2', 42);
$r->zadd('zs1', 1, 'apple');
$r->zadd('zs2', 1, 'banana');

$r->zadd('42', 1, 'apple');

$f = new foo();

print_r($r->zdiff([$f]));
print_r($r->zdiff([42]));
```

We could write specialized logic for when the `zval` is a string, but when that is true the `zend_string_copy` and `zend_string_release` simply increment and decrement the refcount.